Source/drlg_l2.cpp

```diff
@@ -1723,50 +1727,40 @@ bool PlaceMiniSet(const BYTE *miniset, int tmin, int tmax, int cx, int cy, bool
 	return true;
 }
 
-void PlaceMiniSetRandom(const BYTE *miniset, int rndper)
+void PlaceMiniSetRandom(const Miniset &miniset, int rndper)
 {
-	int sw = miniset[0];
-	int sh = miniset[1];
+	int sw = miniset.size.width;
+	int sh = miniset.size.height;
 
 	for (int sy = 0; sy < DMAXY - sh; sy++) {
 		for (int sx = 0; sx < DMAXX - sw; sx++) {
-			bool found = true;
-			int ii = 2;
-			if (sx >= nSx1 && sx <= nSx2 && sy >= nSy1 && sy <= nSy2) {
-				found = false;
-			}
-			for (int yy = 0; yy < sh && found; yy++) {
-				for (int xx = 0; xx < sw && found; xx++) {
-					if (miniset[ii] != 0 && dungeon[xx + sx][yy + sy] != miniset[ii]) {
-						found = false;
-					}
-					if (dflags[xx + sx][yy + sy] != 0) {
-						found = false;
-					}
-					ii++;
+			if (sx >= nSx1 && sx <= nSx2 && sy >= nSy1 && sy <= nSy2)
+				goto next;
+			for (int yy = 0; yy < sh; yy++) {
+				for (int xx = 0; xx < sw; xx++) {
+					if (miniset.search[yy][xx] != 0 && dungeon[xx + sx][yy + sy] != miniset.search[yy][xx])
+						goto next;
+					if (dflags[xx + sx][yy + sy] != 0)
+						goto next;

```

Comment:
can we do this wwithout introducing goto, maybe some sub functions?

Reply:
I'm very partial towards using goto, especially for the purpose of simplifying the control flow of a function. However, I made sure to place it in a separate commit. If you're dead set against goto, just merge everything aside from https://github.com/diasurgical/devilutionX/pull/2728/commits/6f3586fb58c7bbe1a8b99991e65e19a1a64feb1f.

Reply:
It's nice for breaking out of nested loops, but that only really applies to the inner checks (miniset overlaps a pre-existing set, miniset search doesn't match the full pattern). The search code in particular could be reused in PlaceMiniSet if it were a function.